### ISSUE TITLE
Support for --profile argument when deploying to AWS

### ISSUE THREAD

**Issue by flomotlik** [serverless/serverless#1787]
##### Feature Request:

it should be possible to run `serverless deploy --profile production` to use the production AWS profile. At the moment this has to be set through environment variables
##### Benefits:
- being able to easily switch profiles to deploy with

**Comment by kevinaud** [comment]
Just to clarify, could the argument after the --profile option be any named profile that is configured on the AWS cli?

**Comment by flomotlik** [comment]
@kevinaud exactly.

**Comment by kevinaud** [comment]
okay, I'd like to work on this

**Comment by flomotlik** [comment]
Awesome, thanks. I set the labels for it. Its probably best to add it somewhere here: https://github.com/serverless/serverless/blob/990d7c5a089c2047cdd342fca88f6c006c53c4c3/lib/plugins/aws/index.js#L83

**Comment by kevinaud** [comment]
Should setting the profile with environment variables still be supported once this is in place?

**Comment by hassankhan** [comment]
I would hope not, generally we should be using the `~/.aws/credentials` file. Also leaves less room for WTF moments 😄

**Comment by kevinaud** [comment]
So the three possible scenarios as I understand it are:

 1. `--profile` option **isn't** used and the `AWS_PROFILE` environment variable **isn't** set, in which case deploy with the 'default' profile
 2. `--profile` option **isn't** used and the `AWS_PROFILE` environment variable **is** set, in which case use the value of that environment variable as name of the profile to deploy with 
 3. `--profile` option **is** used and, in which case use the value of the `--profile` argument as name of the profile to deploy with regardless of whether or not the `AWS_PROFILE` environment variable is set

**Comment by hassankhan** [comment]
I believe I saw a precedence order [here](https://github.com/serverless/serverless/issues/1801#issuecomment-238600591): 

CLI Option > ENV > serverless.yml

**Comment by kevinaud** [comment]
Is setting the profile in serverless.yml currently supported? I don't see it in the docs https://github.com/serverless/serverless/blob/master/docs/understanding-serverless/serverless-yml.md

**Comment by hassankhan** [comment]
I don't think it is supported **yet**, possibly in the next beta.

**Comment by flomotlik** [comment]
@hassankhan @kevinaud setting the profile will most likely not be supported in the serverless.yml directly.

as @hassankhan already mentioned CLI > ENV

But as CLI always has precedence basically what we can do is if there is a cli option we're setting the profile directly, if no CLI option is given we don't set anything (not even default) and let the AWS SDK deal with it.

**Comment by flomotlik** [comment]
@kevinaud just wanted to ping if you had time to look into it or need any help?

**Comment by kevinaud** [comment]
@flomotlik Sorry I've been in the process of moving back to college (which includes travelling across most of the US for me) but I'm about to submit a pull request for you to check out.

**Comment by MrRoyce** [comment]
To anyone else that's not sure how to set the environment variables for the AWS profile during deployment:

$ AWS_PROFILE=<put the name of the aws profile here> serverless deploy

**Comment by kennu** [comment]
I just tested Serverless 1.0.0-beta2 and was very confused how to choose the AWS profile. Previously Serveless 0.x has asked for it (with a menu) and stored it in admin.env, which we have stored in Git, so that the project is always automatically deployed to the right profile. I hope this functionality can be restored in 1.0.

**Comment by flomotlik** [comment]
@kevinaud submitted the PR now btw: #1905 great job and will be reviewed and hopefully merged soon.

@kennu so you want to hardcode the profile name in the serverless.yml? Our current approach would be to follow AWS defaults and use `--profile` if its available, AWS_PROFILE if that is available or the default profile in the .aws/credentials file.

Are you the only person deploying those services so you want to hardcode them, or is the hardcoded profile used across a few people in the team?

**Comment by kennu** [comment]
In my case (workflow with Serverless 0.5 so far)
- I deploy to 17 different AWS accounts, so each project needs some way to remember the profile or things get difficult
- AWS profile names are standardized between team members so that admin.env can be shared in Git

Until now, we could trust that "sls xxx deploy" does the right thing automatically, so it's why I'm missing that feature. Having to prefix commands with AWS_PROFILE=xxx manually is fairly inconvenient and error-prone.

**Comment by flomotlik** [comment]
Ok thats good to know. I can see something like

```yaml
provider:
  profile: production
```

for example. If you provide it via `--profile` you can override that setting, but by default it should work and deploy properly.

**Comment by kennu** [comment]
@flomotlik I think that would work well. In fact I tried put `profile: xxx` in serverless.yml, because GitHub search found some test file indicating it might work. And then was disappointed. :-)

It is also a larger problem, because it would be also desirable for AWS CLI to use a specific profile when working under a project. But that's another story and with Serverless, there's generally not that much need to use AWS CLI every day.

**Comment by kennu** [comment]
This also reminds me of another discussion about overriding AWS profiles for each deployment target. We have many cases where production stage is in a separate AWS account. It would be nice to have `sls deploy -s prod` deploy to the right profile automatically.

**Comment by rowanu** [comment]
I'm not sure @flomotlik's example would work - wouldn't that set your profile for your entire provider (i.e. AWS)? I think profile needs to be set per stage for the reasons @kennu mentioned (i.e. deploying different stages to different AWS accounts - it's just best practice).

While the `--profile` flag is a step in the right direction, I would really like to set a profile to use for deployment in configuration (probably `serverless.env.yml`?) I'm thinking something like this:

```
stages:
  dev:
    vars:
      profile: my-dev-profile
    regions:
      ap-southeast-2:
        vars:
 prod:
    vars:
      profile: my-prod-profile
    regions:
      ap-southeast-2:
        vars:
```

Where `my-dev-profile` and `my-prod-profile` just happen to be different AWS accounts.

Does that make sense, or am I missing something? I saw [this test](https://github.com/serverless/serverless/blob/master/lib/plugins/aws/tests/index.js#L87) and thought it would "just work" but it doesn't seem to be the case - any `sls` commands just bail if I don't have my `AWS_PROFILE` env var set...

**Comment by rowanu** [comment]
After a bit more investigation, I think the issue I described above is not with Serverless - it seems to be getting the profile from config correctly. It seems like the AWS SDK is not aware of the setting... Sorry for the issue hijack :neutral_face:

**Comment by flomotlik** [comment]
@rowanu you can do per stage setting with nesting of environment variables as described in #1834

**Comment by rowanu** [comment]
@flomotlik the problem is the variables weren't honoured by the SDK. Please see my (recently updated) bug #1940 and PR #1945.

**Comment by flomotlik** [comment]
@rowanu we're introducing the new variable system in the next few days which will make things like this easier going forward

**Comment by mwawrusch** [comment]
Being a bit needy here, but can we have a beta.3 drop with the profiles per stage ready?

**Comment by rowanu** [comment]
@mwawrusch this issue is not for profiles per stage, you want #1940 and #2015.

**Comment by kennu** [comment]
Btw, Serverless 1.0-RC2 still *silently ignores* the --profile option. No error message, just deploys to the wrong profile (default). I got bitten by this just now once again. I think it should at least show an error message and abort when you specify an unsupported option.

**Comment by kevinold** [comment]
I'm seeing an issue where serverless doesn't honor a role based setup of credentials:

```
[default]
aws_access_key_id=AAAAAAAAAAAA
aws_secret_access_key=XXXXXXXXXXXXXXXXXXXXXXX

  [sandbox-user]
  role_arn = arn:aws:iam::AAAAAA:role/FullAdmin
  source_profile = default
  mfa_serial = arn:aws:iam::ZZZZZZZ:mfa/kold
```

I've added the sandbox-user profile to serverless.yaml and get `Error: ServerlessError: ServerlessError: AWS provider credentials not found. You can find more info on how to set up provider credentials in our docs here: https://git.io/viZAC` when I run it.  I've also tried the env vars but nothing works.

**Comment by rowanu** [comment]
@kennu @kevinaud guys, what you've both described are valid issues, but neither belongs on this issue. Please open a new issue or raise it in [the forum](http://forum.serverless.com/) if you want it addressed.

**Comment by jthomerson** [comment]
I agree that **if** support is added to specify the profile in the `serverless.yml` file like it was in `admin.env` pre 1.0, it must support the configuration of different profiles per stage, and a default profile. That supports the (common) usecase where you deploy developers' builds to one account, integration builds to another, and production to yet another.

**Comment by andymac4182** [comment]
Is this not supported with the variables work and loading a variable from eg. `${opt:profile}`?

**Comment by flomotlik** [comment]
Yup as @andymac4182 said this should be already done and was expanded on recently in https://github.com/serverless/serverless/pull/2229

Documented here: https://serverless.com/framework/docs/providers/aws/setup/

While this doesn't have `--profile` directly built into as an option I think its a cleaner implementation because its more specific on the users end. We might add this more directly through options in the future though, should be easy with that new implementation.

As I think this can be closed now I'll close.

**Comment by RobGruhl** [comment]
Bad doc link?  I think this is the new one: https://serverless.com/framework/docs/providers/aws/guide/credentials/

**Comment by DavidWells** [comment]
This flag would be great for numerous reasons.

Can we revisit this? 

I can implement

**Comment by pmuens** [comment]
@nikgraf @DavidWells that's what you need for the dashboard, right?

**Comment by skilledDeveloper** [comment]
Like @kevinold mentioned, I have a role-based AWS setup and I'm not sure how I can deploy to that account. 
Setting the env variable `AWS_PROFILE=my-role-based-profile` doesn't help as I get `ServerlessError: ServerlessError: AWS provider credentials not found.`

It used to work in the 0.5.x because every stage was tied to a specific profile.
Any workaround?

**Comment by kevinold** [comment]
@skilledDeveloper I ended up creating another user in my role-based account and used that user as my profile and all worked well.

**Comment by jokeyrhyme** [comment]
From AWS documentation, the environment variable is `AWS_DEFAULT_PROFILE`

https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-getting-started.html#cli-environment

Is the `AWS_PROFILE` variable discussed here intentionally different?

**Comment by adonisv79** [comment]
I wonder why this does not exist... this would be great to have as we devs may need to work on multiple AWS accounts (which I do now). AWS CLI allows us to configure multiple accounts and even set the default regions for each. why not allow us to load default or customized profiles using them?

serverless deploy --profile [default | mycustomprofile]

**Comment by nappa32** [comment]
Currently I use a setup where I have 24 AWS profiles configured. There is one AWS account, the rest are ROLE based profiles. The only way I can make serverless work is by creating STS sessions and populating ENV variables. This means the values will expire every 1 hour.

serverless should support role based aws cli profiles.

**Comment by jokeyrhyme** [comment]
@nappa32 that might be better served with an `--assume-role` CLI parameter, maybe?

**Comment by devtanc** [comment]
In the meantime...I wrote this to solve the issue for me: [aws-profile-switcher](https://www.npmjs.com/package/aws-profile-switcher)

**Comment by jelling** [comment]
FYI, the docs specify this feature as being implemented but it is not yet released. The feature is in v1.15 but v1.14 is the latest on NPM. 

Hopefully this will save someone else some time.

**Comment by pmuens** [comment]
> FYI, the docs specify this feature as being implemented but it is not yet released. The feature is in v1.15 but v1.14 is the latest on NPM.

Thanks for mentioning @jelling 👍 

Yes, we're releasing this with `v1.15` on Wednesday. We're also looking into ways to publish the docs for specific versions rather than building it from `master`.

**Comment by skilledDeveloper** [comment]
Can anyone tell me if this new feature supports role-based profiles?
(as described here: https://github.com/serverless/serverless/issues/1787#issuecomment-268129493)

**Comment by pmuens** [comment]
@skilledDeveloper haven't tested that yet 🤔 

The code is pretty minimalistic: https://github.com/serverless/serverless/pull/3701/files#diff-271ef5d492727880dcac1b94a86c878d

**Comment by landonreed** [comment]
If you're looking for the documentation for on-the-fly switching profiles, you can find it here: https://serverless.com/framework/docs/providers/aws/guide/credentials/#using-aws-profiles

`serverless deploy --aws-profile devProfile `

**Comment by dpwrussell** [comment]
Should this be reopened (or a new issue raised) to cope with role-based profiles? I just started using serverless and was stopped in my tracks on hello-world because I need to use a profile configured in `~/.aws/config` to enable deployment.

E.g.

```
[profile sudo]
role_arn = arn:aws:iam::************:role/SUDO
source_profile = default
region = us-east-1
```

**Comment by didac-wh** [comment]
Is there any elegant solution for those of us who deploy from an AWS instance?
When deploying directly from AWS the param "profile" is not recognized, not even with the "default" keyword.

For now I've removed the "profile: [whatever]" from my serverless.yml declaration and when I need to deploy from my local environment I specify the profile directly from the command line like this:
`sls deploy -v --stage stg --aws-profile [my-awesome-profile]`

**Comment by funkswing** [comment]
@dpwrussell You can use role-based AWS profiles with the `assume-role` tool:
https://github.com/remind101/assume-role

**Comment by joeymiller** [comment]
Can confirm that setting the AWS_SDK_LOAD_CONFIG env var to a truthy value works. 

`export AWS_SDK_LOAD_CONFIG=1`

**Comment by akash-rajput** [comment]
How about option by yourself

```
provider:
  name: aws
  runtime: nodejs12.x
  profile: ${opt:profile, 'dev'}
  stage: ${opt:stage, 'dev'}
```
